frame/paged-list/src/paged_list.rs

```diff
@@ -0,0 +1,530 @@
+		meta: StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage>,
+		drain: bool,
+	) -> Self {
+		let mut page = Page::<Value>::from_storage::<Prefix, Hasher>(meta.first_page);
+
+		if let Some(ref mut page) = page {
+			page.next = meta.first_value;
+		}
+
+		Self { page, drain, meta }
+	}
+}
+
+impl<Prefix, Hasher, Value, ValuesPerPage> Iterator
+	for StoragePagedListIterator<Prefix, Hasher, Value, ValuesPerPage>
+where
+	Prefix: StorageInstance,
+	Value: FullCodec + Clone,
+	Hasher: StorageHasher,
+	ValuesPerPage: Get<u32>,
+{
+	type Item = Value;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		let page = self.page.as_mut()?;
+
+		if let Some(val) = page.next() {
+			if self.drain {
+				self.meta.first_value.saturating_inc();
+				self.meta.store()
+			}
+			return Some(val)
+		}
+		if self.drain {
+			// page is empty
+			page.delete::<Prefix, Hasher>();
+			self.meta.first_value = 0;
+			self.meta.first_page.saturating_inc();
+			self.meta.store();
+		}
+
+		let Some(mut page) = Page::from_storage::<Prefix, Hasher>(page.index.saturating_add(1)) else {
+			self.page = None;
+			if self.drain {
+				self.meta.reset();
+			}
+			return None;
+		};
+
+		if let Some(val) = page.next() {
+			self.page = Some(page);
+			if self.drain {
+				self.meta.first_value.saturating_inc();
+				self.meta.store();
+			}
+			return Some(val)
+		}
+
+		defensive!("StoragePagedListIterator: empty pages do not exist: storage corrupted");
+		// Put the storage back into a consistent state.
+		self.meta.reset();
+		self.page = None;
+		None
+	}
+}
+
+impl<Prefix, Hasher, Value, ValuesPerPage, MaxPages> frame_support::storage::StorageList<Value>
+	for StoragePagedList<Prefix, Hasher, Value, ValuesPerPage, MaxPages>
+where
+	Prefix: StorageInstance,
+	Value: FullCodec + Clone,
+	Hasher: StorageHasher,
+	ValuesPerPage: Get<u32>,
+	MaxPages: Get<Option<u32>>,
+{
+	type Iterator = StoragePagedListIterator<Prefix, Hasher, Value, ValuesPerPage>;
+	type Appendix = StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage>;
+
+	fn iter() -> Self::Iterator {
+		StoragePagedListIterator::from_meta(Self::read_meta(), false)
+	}
+
+	fn drain() -> Self::Iterator {
+		StoragePagedListIterator::from_meta(Self::read_meta(), true)
+	}
+
+	fn appendix() -> Self::Appendix {
+		Self::appendix()
+	}
+}
+
+impl<Prefix, Hasher, Value, ValuesPerPage, MaxPages>
+	StoragePagedList<Prefix, Hasher, Value, ValuesPerPage, MaxPages>
+where
+	Prefix: StorageInstance,
+	Value: FullCodec + Clone,
+	Hasher: StorageHasher,
+	ValuesPerPage: Get<u32>,
+	MaxPages: Get<Option<u32>>,
+{
+	fn read_meta() -> StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage> {
+		// Use default here to not require a setup migration.
+		StoragePagedListMeta::from_storage().unwrap_or_default()
+	}
+
+	/// Provides a fast append iterator.
+	///
+	/// The list should not be modified while appending. Also don't call it recursively.
+	fn appendix() -> StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage> {
```

Comment:
Honestly `appendix` seems like kind of a weird name to me; reminds me either [of an actual appendix I have in my belly](https://en.wikipedia.org/wiki/Appendix_(anatomy)) or of an extra section at the end of a book. (: Maybe, hm, `appender` would be more appropriate?

Reply:
>or of an extra section at the end of a book

This was the idea :laughing: 
>[Addendum](https://en.wikipedia.org/wiki/Addendum), an addition made to a document by its author after its initial printing or publication

but i can also go for `appender`, probably more idiomatic. Rust does not seem to have a standard name for this. In C++ it would be `back_insert_iterator`…

Reply:
Renamed to `Appender`.